File: helper/internal/client/agent.go
// Agent-mode calls to /api/agent/*, authenticated with a node's persistent
// agent token rather than the one-shot pairing code ReportSpec (client.go)
// uses — this credential authenticates a process that runs indefinitely,
// not a single report.
package client

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// AgentJob is what the backend hands the agent to run — mirrors
// agentJobDto in server/src/routes/agent.js.
type AgentJob struct {
	Version   int       `json:"version"`
	JobID     string    `json:"jobId"`
	NodeID    string    `json:"nodeId"`
	Attempt   int       `json:"attempt"`
	IssuedAt  time.Time `json:"issuedAt"`
	ExpiresAt time.Time `json:"expiresAt"`
	Seed      uint32    `json:"seed"`
	InputHash string    `json:"inputHash"`
	Execution struct {
		WorkloadID      string            `json:"workloadId"`
		DockerImage     string            `json:"dockerImage"`
		ModelID         string            `json:"modelId"`
		CacheKey        string            `json:"cacheKey"`
		GPUsNeeded      int               `json:"gpusNeeded"`
		MaxRuntimeHours float64           `json:"maxRuntimeHours"`
		EnvVars         map[string]string `json:"envVars"`
	} `json:"execution"`
	Inputs []JobInput      `json:"inputs"`
	Result JobResultPolicy `json:"result"`
}

type JobInput struct {
	ID          string `json:"id"`
	Filename    string `json:"filename"`
	ContentType string `json:"contentType"`
	ByteSize    int64  `json:"byteSize"`
	SHA256      string `json:"sha256"`
}

type JobResultPolicy struct {
	Schema              string   `json:"schema"`
	Required            bool     `json:"required"`
	MaxBytes            int64    `json:"maxBytes"`
	AllowedContentTypes []string `json:"allowedContentTypes"`
}

type Benchmark struct {
	Version      string  `json:"version"`
	GPUScore     float64 `json:"gpuScore,omitempty"`
	DownloadMbps float64 `json:"downloadMbps,omitempty"`
	UploadMbps   float64 `json:"uploadMbps,omitempty"`
}

type Capabilities struct {
	CachedModels []string  `json:"cachedModels"`
	Benchmark    Benchmark `json:"benchmark"`
}

type signedJob struct {
	Manifest  json.RawMessage `json:"manifest"`
	Signature string          `json:"signature"`
}

type LogLine struct {
	Level string `json:"level"`
	Msg   string `json:"msg"`
}

type HeartbeatSample struct {
	GPUUsagePct float64 `json:"gpuUsagePct"`
	VRAMUsedGB  float64 `json:"vramUsedGb"`
}

type AgentClient struct {
	apiBase string
	token   string
	http    *http.Client
}

func NewAgentClient(apiBase, token string) *AgentClient {
	return &AgentClient{apiBase: apiBase, token: token, http: &http.Client{Timeout: 15 * time.Second}}
}

func (c *AgentClient) BenchmarkNetwork() (downloadMbps, uploadMbps float64) {
	const size = 256 * 1024
	getURL := c.apiBase + "/api/agent/benchmark"
	req, _ := http.NewRequest(http.MethodGet, getURL, nil)
	req.Header.Set("Authorization", "Bearer "+c.token)
	start := time.Now()
	resp, err := c.http.Do(req)
	if err == nil {
		data, readErr := io.ReadAll(io.LimitReader(resp.Body, size+1))
		resp.Body.Close()
		if readErr == nil && resp.StatusCode < 300 && len(data) == size {
			downloadMbps = float64(size*8) / time.Since(start).Seconds() / 1_000_000
		}
	}
	payload := bytes.Repeat([]byte{0x5a}, size)
	req, _ = http.NewRequest(http.MethodPost, getURL, bytes.NewReader(payload))
	req.Header.Set("Authorization", "Bearer "+c.token)
	req.Header.Set("Content-Type", "application/octet-stream")
	start = time.Now()
	resp, err = c.http.Do(req)
	if err == nil {
		io.Copy(io.Discard, resp.Body)
		resp.Body.Close()
		if resp.StatusCode < 300 {
			uploadMbps = float64(size*8) / time.Since(start).Seconds() / 1_000_000
		}
	}
	return
}

func (c *AgentClient) do(method, path string, body, out interface{}) error {
	var reader io.Reader
	if body != nil {
		payload, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reader = bytes.NewReader(payload)
	}

	url := c.apiBase + path
	req, err := http.NewRequest(method, url, reader)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return &connError{url: url, cause: err}
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		snippet, _ := io.ReadAll(io.LimitReader(resp.Body, 300))
		return &statusError{url: url, status: resp.StatusCode, body: strings.TrimSpace(string(snippet))}
	}
	if out != nil {
		return json.NewDecoder(resp.Body).Decode(out)
	}
	return nil
}

// Heartbeat reports liveness and returns the node's current job, if any —
// one combined call rather than two independent poll loops for what is
// conceptually one "what should I be doing" question.
func (c *AgentClient) Heartbeat(capabilities Capabilities) (*AgentJob, error) {
	var out struct {
		Data struct {
			Job *signedJob `json:"job"`
		} `json:"data"`
	}
	if err := c.do(http.MethodPost, "/api/agent/heartbeat", struct {
		Capabilities Capabilities `json:"capabilities"`
	}{capabilities}, &out); err != nil {
		return nil, err
	}
	if out.Data.Job == nil {
		return nil, nil
	}
	return c.verifyJob(out.Data.Job)
}

func (c *AgentClient) verifyJob(signed *signedJob) (*AgentJob, error) {
	provided, err := base64.RawURLEncoding.DecodeString(signed.Signature)
	if err != nil {
		return nil, errors.New("job manifest has an invalid signature encoding")
	}
	mac := hmac.New(sha256.New, []byte(c.token))
	mac.Write(signed.Manifest)
	if !hmac.Equal(provided, mac.Sum(nil)) {
		return nil, errors.New("job manifest signature verification failed")
	}
	var job AgentJob
	if err := json.Unmarshal(signed.Manifest, &job); err != nil {
		return nil, fmt.Errorf("job manifest is invalid: %w", err)
	}
	nodeID, _, ok := strings.Cut(c.token, ".")
	if !ok || job.NodeID != nodeID {
		return nil, errors.New("job manifest was issued for a different node")
	}
	now := time.Now()
	if job.ExpiresAt.Before(now) || job.ExpiresAt.After(now.Add(2*time.Minute)) {
		return nil, errors.New("job manifest is expired or has an invalid lifetime")
	}
	if job.Version != 2 || job.JobID == "" || job.Execution.WorkloadID == "" || job.Execution.DockerImage == "" ||
		job.Execution.GPUsNeeded < 1 || job.Execution.GPUsNeeded > 16 ||
		job.Execution.MaxRuntimeHours <= 0 || job.Execution.MaxRuntimeHours > 72 ||
		job.InputHash == "" || job.Result.Schema == "" {
		return nil, errors.New("job manifest failed validation")
	}
	return &job, nil
}

func (c *AgentClient) DownloadInput(jobID string, input JobInput) ([]byte, error) {
	url := c.apiBase + "/api/agent/job/" + jobID + "/input/" + input.ID
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	resp, err := c.http.Do(req)
	if err != nil {
		return nil, &connError{url: url, cause: err}
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("input download returned status %d", resp.StatusCode)
	}
	data, err := io.ReadAll(io.LimitReader(resp.Body, input.ByteSize+1))
	if err != nil {
		return nil, err
	}
	if int64(len(data)) != input.ByteSize {
		return nil, errors.New("downloaded input size does not match manifest")
	}
	digest := sha256.Sum256(data)
	if fmt.Sprintf("%x", digest) != input.SHA256 {
		return nil, errors.New("downloaded input SHA-256 does not match manifest")
	}
	return data, nil
}

func (c *AgentClient) ClaimJob(jobID string) error {
	return c.do(http.MethodPost, "/api/agent/job/"+jobID+"/claim", struct{}{}, nil)
}

func (c *AgentClient) PostHeartbeats(jobID string, samples []HeartbeatSample) error {
	body := struct {
		Samples []HeartbeatSample `json:"samples"`
	}{Samples: samples}
	return c.do(http.MethodPost, "/api/agent/job/"+jobID+"/heartbeats", body, nil)
}

func (c *AgentClient) PostLogs(jobID string, lines []LogLine) error {
	body := struct {
		Lines []LogLine `json:"lines"`
	}{Lines: lines}
	return c.do(http.MethodPost, "/api/agent/job/"+jobID+"/logs", body, nil)
}

// UploadArtifact sends a job's output file, read off the container's
// mounted /output directory, as a raw body — not JSON, so this bypasses
// do() and sets Content-Type to the artifact's own type directly.
func (c *AgentClient) UploadArtifact(jobID, contentType string, data []byte) error {
	url := c.apiBase + "/api/agent/job/" + jobID + "/artifact"
	req, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+c.token)
	// Keep the transport opaque so the API's global JSON parser cannot alter
	// structured result bytes before their SHA-256 is verified.
	req.Header.Set("Content-Type", "application/octet-stream")
	req.Header.Set("X-Decompute-Result-Content-Type", contentType)
	digest := sha256.Sum256(data)
	req.Header.Set("X-Decompute-Sha256", fmt.Sprintf("%x", digest))

	resp, err := c.http.Do(req)
	if err != nil {
		return &connError{url: url, cause: err}
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		snippet, _ := io.ReadAll(io.LimitReader(resp.Body, 300))
		return &statusError{url: url, status: resp.StatusCode, body: strings.TrimSpace(string(snippet))}
	}
	return nil
}

func (c *AgentClient) CompleteJob(jobID, status, errorMessage string) error {
	body := struct {
		Status       string `json:"status"`
		ErrorMessage string `json:"errorMessage,omitempty"`
	}{Status: status, ErrorMessage: errorMessage}
	return c.do(http.MethodPost, "/api/agent/job/"+jobID+"/complete", body, nil)
}

type connError struct {
	url   string
	cause error
}

func (e *connError) Error() string {
	return "couldn't reach Decompute at " + e.url + "\n  cause: " + e.cause.Error() + "\n  " + hintFor(e.cause)
}
func (e *connError) Unwrap() error { return e.cause }

type statusError struct {
	url    string
	status int
	body   string
}

func (e *statusError) Error() string {
	msg := "Decompute returned an error (status " + strconv.Itoa(e.status) + ") from " + e.url
	if e.body != "" {
		msg += "\n  " + e.body
	}
	return msg
}

File: server/src/routes/agent.js
import express, { Router } from "express";
import { pool, query } from "../db.js";
import { requireAgentAuth } from "../middleware/requireAgentAuth.js";
import { getWorkload } from "../lib/workloadCatalog.js";
import { signJobManifest } from "../lib/jobManifest.js";
import { validateArtifact } from "../lib/resultValidation.js";
import { incidentModeEnabled } from "../lib/incidentMode.js";
import { settleJob } from "../lib/jobSettlement.js";

export const agentRouter = Router();

const BENCHMARK_BYTES = 256 * 1024;
const benchmarkPayload = Buffer.alloc(BENCHMARK_BYTES, 0x5a);
agentRouter.get("/benchmark", requireAgentAuth, (req, res) => {
  res.setHeader("Content-Type", "application/octet-stream");
  res.setHeader("Cache-Control", "no-store");
  res.send(benchmarkPayload);
});
agentRouter.post("/benchmark", requireAgentAuth, express.raw({ type: "application/octet-stream", limit: BENCHMARK_BYTES }), (req, res) => {
  if (!Buffer.isBuffer(req.body) || req.body.length !== BENCHMARK_BYTES) return res.status(400).json({ error: "Invalid benchmark payload" });
  res.json({ ok: true });
});

// Combined liveness ping + "what should I be doing" — one poll loop covers
// both, rather than a resident agent needing two independent intervals for
// what is conceptually one question. Called every ~15s by the agent.
agentRouter.post("/heartbeat", requireAgentAuth, async (req, res) => {
  const cachedModels = Array.isArray(req.body?.capabilities?.cachedModels)
    ? [...new Set(req.body.capabilities.cachedModels.map(String).filter((v) => v.length <= 300))].slice(0, 100)
    : [];
  const benchmark = req.body?.capabilities?.benchmark || {};
  const numberOrNull = (value, max) => {
    const n = Number(value);
    return Number.isFinite(n) && n >= 0 && n <= max ? n : null;
  };
  await query(`UPDATE nodes SET last_seen_at = now(), cached_models = $2,
      gpu_benchmark_score = COALESCE($3, gpu_benchmark_score),
      network_download_mbps = COALESCE($4, network_download_mbps),
      network_upload_mbps = COALESCE($5, network_upload_mbps),
      benchmark_version = COALESCE($6, benchmark_version), capabilities_reported_at = now()
    WHERE id = $1`, [req.nodeId, JSON.stringify(cachedModels),
      numberOrNull(benchmark.gpuScore, 1e12), numberOrNull(benchmark.downloadMbps, 1e7),
      numberOrNull(benchmark.uploadMbps, 1e7), benchmark.version ? String(benchmark.version).slice(0, 50) : null]);

  const { rows } = await query(
    `SELECT * FROM jobs WHERE node_id = $1 AND status IN ('pending','running') ORDER BY created_at ASC LIMIT 1`,
    [req.nodeId]
  );
  if (!rows[0]) return res.json({ data: { job: null, incidentMode: incidentModeEnabled() } });
  const job = rows[0];
  const inputs = (await query(
    `SELECT id::text, filename, content_type, sha256, byte_size FROM job_inputs WHERE job_id = $1 ORDER BY id`, [job.id]
  )).rows;
  const workload = getWorkload(job.workload_id);
  if (!workload) return res.status(409).json({ error: "Assigned workload is no longer enabled" });
  const signed = signJobManifest(job, req.nodeId, req.agentToken, inputs, workload);
  await query(`UPDATE jobs SET manifest_hash = $1 WHERE id = $2 AND manifest_hash IS NULL`, [signed.manifestHash, job.id]);
  res.json({ data: { job: { manifest: signed.manifest, signature: signed.signature }, incidentMode: incidentModeEnabled() } });
});

// Transitions pending -> running. 409 if another poll already claimed it,
// the job doesn't belong to this node, or it's not in a claimable state.
agentRouter.post("/job/:id/claim", requireAgentAuth, async (req, res) => {
  const { rows } = await query(
    `UPDATE jobs SET status = 'running', started_at = now()
     WHERE id = $1 AND node_id = $2 AND status = 'pending'
     RETURNING id, started_at`,
    [req.params.id, req.nodeId]
  );
  if (!rows[0]) return res.status(409).json({ error: "Job is not claimable" });
  res.json({ data: { id: rows[0].id, startedAt: rows[0].started_at } });
});

// job_heartbeats/job_logs have no owner column of their own — every write
// below re-checks node_id and status='running' so a stale or misdirected
// agent can't write into a job it doesn't currently hold.
async function ownsRunningJob(jobId, nodeId) {
  const { rows } = await query(
    `SELECT 1 FROM jobs WHERE id = $1 AND node_id = $2 AND status = 'running'`,
    [jobId, nodeId]
  );
  return !!rows[0];
}

agentRouter.post("/job/:id/heartbeats", requireAgentAuth, async (req, res) => {
  if (!(await ownsRunningJob(req.params.id, req.nodeId))) return res.status(409).json({ error: "Job is not running on this node" });

  const samples = Array.isArray(req.body?.samples) ? req.body.samples.slice(0, 100) : [];
  for (const s of samples) {
    const gpu = Number(s?.gpuUsagePct);
    const vram = Number(s?.vramUsedGb);
    await query(
      `INSERT INTO job_heartbeats (job_id, gpu_usage_pct, vram_used_gb) VALUES ($1, $2, $3)`,
      [req.params.id, Number.isFinite(gpu) ? gpu : null, Number.isFinite(vram) ? vram : null]
    );
  }
  res.json({ ok: true });
});

const ARTIFACT_MAX_BYTES = 15 * 1024 * 1024;

agentRouter.get("/job/:id/input/:inputId", requireAgentAuth, async (req, res) => {
  const { rows } = await query(
    `SELECT i.filename, i.content_type, i.sha256, i.data FROM job_inputs i
     JOIN jobs j ON j.id = i.job_id
     WHERE i.id = $1 AND j.id = $2 AND j.node_id = $3 AND j.status IN ('pending','running')`,
    [req.params.inputId, req.params.id, req.nodeId]
  );
  if (!rows[0]) return res.status(404).json({ error: "Input not found" });
  res.setHeader("Content-Type", rows[0].content_type);
  res.setHeader("X-Decompute-Sha256", rows[0].sha256);
  res.setHeader("Content-Disposition", `attachment; filename="${rows[0].filename.replace(/["\\]/g, "_")}"`);
  res.setHeader("Cache-Control", "private, no-store");
  res.send(rows[0].data);
});

// The agent uploads a job's output file (read off the container's mounted
// output directory) after the run finishes but before /complete — the job
// is still 'running' at that point, so this reuses the same ownership
// guard as heartbeats/logs. Raw bytes, not JSON: express.raw() below caps
// the body at ARTIFACT_MAX_BYTES so an oversized file 413s instead of
// filling memory.
agentRouter.post("/job/:id/artifact", requireAgentAuth, express.raw({ type: "*/*", limit: ARTIFACT_MAX_BYTES }), async (req, res) => {
  if (!(await ownsRunningJob(req.params.id, req.nodeId))) return res.status(409).json({ error: "Job is not running on this node" });

  const job = (await query(`SELECT workload_id, result_schema FROM jobs WHERE id = $1`, [req.params.id])).rows[0];
  const workload = getWorkload(job.workload_id);
  const contentType = String(req.headers["x-decompute-result-content-type"] || req.headers["content-type"] || "application/octet-stream").split(";")[0];
  try {
    const result = validateArtifact({ schema: job.result_schema, contentType, data: req.body,
      claimedHash: String(req.headers["x-decompute-sha256"] || ""),
      allowedContentTypes: workload?.allowedContentTypes || [] });
    await query(`INSERT INTO job_artifacts
      (job_id, content_type, data, sha256, byte_size, validated, validation_metadata)
      VALUES ($1,$2,$3,$4,$5,true,$6)`,
      [req.params.id, contentType, req.body, result.sha256, result.byteSize, JSON.stringify(result.metadata)]);
    await query(`UPDATE jobs SET result_validated = true, result_hash = $1, result_metadata = $2 WHERE id = $3`,
      [result.sha256, JSON.stringify(result.metadata), req.params.id]);
    res.json({ ok: true, data: result });
  } catch (err) {
    res.status(422).json({ error: err.message });
  }
});

const LOG_LEVELS = ["INFO", "WARN", "ERROR", "DEBUG"];

agentRouter.post("/job/:id/logs", requireAgentAuth, async (req, res) => {
  if (!(await ownsRunningJob(req.params.id, req.nodeId))) return res.status(409).json({ error: "Job is not running on this node" });

  const lines = Array.isArray(req.body?.lines) ? req.body.lines.slice(0, 200) : [];
  for (const l of lines) {
    const level = LOG_LEVELS.includes(l?.level) ? l.level : "INFO";
    const msg = String(l?.msg || "").slice(0, 4000);
    if (!msg) continue;
    await query(`INSERT INTO job_logs (job_id, level, msg) VALUES ($1, $2, $3)`, [req.params.id, level, msg]);
  }
  res.json({ ok: true });
});

// Settles the job. Failed jobs are fully refunded — no payout for
// incomplete work. Completed jobs bill for actual elapsed time (capped at
// what was paid for), refund the unused portion, and release the billed
// amount split between the node owner and platform_revenue. Two
// postTransaction calls, not one: userBalanceDelta only updates one user's
// cached balance per call, and the refund (renter) and release (node
// owner) are two different users.
agentRouter.post("/job/:id/complete", requireAgentAuth, async (req, res) => {
  const status = req.body?.status === "failed" ? "failed" : "done";
  const errorMessage = req.body?.errorMessage ? String(req.body.errorMessage).slice(0, 4000) : null;

  const client = await pool.connect();
  try {
    await client.query("BEGIN");

    const jobRes = await client.query(
      `SELECT * FROM jobs WHERE id = $1 AND node_id = $2 AND status = 'running' FOR UPDATE`,
      [req.params.id, req.nodeId]
    );
    const job = jobRes.rows[0];
    if (!job) {
      await client.query("ROLLBACK");
      return res.status(409).json({ error: "Job is not running on this node" });
    }

    if (status === "done" && job.result_schema && !job.result_validated) {
      await client.query("ROLLBACK");
      return res.status(422).json({ error: "A validated result artifact is required before successful completion" });
    }

    if (errorMessage) {
      await client.query(`INSERT INTO job_logs (job_id, level, msg) VALUES ($1, 'ERROR', $2)`, [job.id, errorMessage]);
    }

    if (incidentModeEnabled()) {
      // Record the outcome but make no ledger movement. The immutable escrow
      // hold remains balanced in the database until the reconciliation sweep
      // runs after incident mode is removed.
      await client.query(
        `UPDATE jobs SET status = $1, completed_at = now(), failure_reason = $2,
           settlement_state = 'held' WHERE id = $3`,
        [status, status === "failed" ? (errorMessage || "Provider reported failure") : null, job.id]
      );
      await client.query(`DELETE FROM job_inputs WHERE job_id = $1`, [job.id]);
      await client.query("COMMIT");
      return res.status(202).json({ ok: true, settlementDeferred: true });
    }

    await settleJob(client, { ...job, completed_at: new Date() }, status);
    await client.query("COMMIT");
    return res.json({ ok: true });

    /* istanbul ignore next -- legacy settlement implementation retained
       temporarily below for migration readability; execution returns above. */
    if (errorMessage) {
      await client.query(`INSERT INTO job_logs (job_id, level, msg) VALUES ($1, 'ERROR', $2)`, [job.id, errorMessage]);
    }

    const total = Number(job.total_usd);
    const renterAccountId = await getUserAccountId(client, job.user_id);
    const platformEscrowId = await getPlatformAccountId(client, "platform_escrow");
    let settlementTxnId;
    // Failed jobs are fully refunded — nothing billed. Overwritten below
    // for a successful completion, where these become the real, prorated
    // amount actually kept (as opposed to total_usd, the original escrow
    // hold for the job's full max_runtime_hours).
    let billedSubtotal = 0, billedFee = 0, billedTotal = 0;

    if (status === "failed") {
      settlementTxnId = await postTransaction(client, {
        type: "job_refund",
        referenceType: "job",
        referenceId: job.id,
        lines: [
          { accountId: platformEscrowId, amount: total },
          { accountId: renterAccountId, amount: -total },
        ],
        userBalanceDelta: { userId: job.user_id, amount: total },
      });
    } else {
      const elapsedHours = Math.max(0, (Date.now() - new Date(job.started_at).getTime()) / 3_600_000);
      const billedHours = Math.min(Number(job.max_runtime_hours), elapsedHours);
      billedSubtotal = Math.round(Number(job.price_per_hour) * billedHours * 100) / 100;
      billedFee = Math.round(billedSubtotal * SERVICE_FEE_RATE * 100) / 100;
      billedTotal = Math.round((billedSubtotal + billedFee) * 100) / 100;
      const unusedRefund = Math.round((total - billedTotal) * 100) / 100;

      if (unusedRefund > 0) {
        await postTransaction(client, {
          type: "job_refund",
          referenceType: "job",
          referenceId: job.id,
          lines: [
            { accountId: platformEscrowId, amount: unusedRefund },
            { accountId: renterAccountId, amount: -unusedRefund },
          ],
          userBalanceDelta: { userId: job.user_id, amount: unusedRefund },
        });
      }

      // owner_id can only be null for the pre-agent seed catalog, which
      // can never actually reach a running job (no agent_token_hash means
      // no liveness heartbeat, so it can never be matched) — defensive,
      // not an expected path.
      const nodeRes = await client.query(`SELECT owner_id FROM nodes WHERE id = $1`, [job.node_id]);
      const ownerId = nodeRes.rows[0]?.owner_id;
      const platformRevenueId = await getPlatformAccountId(client, "platform_revenue");

      if (ownerId) {
        const ownerAccountId = await getUserAccountId(client, ownerId);
        settlementTxnId = await postTransaction(client, {
          type: "job_escrow_release",
          referenceType: "job",
          referenceId: job.id,
          lines: [
            { accountId: platformEscrowId, amount: billedTotal },
            { accountId: ownerAccountId, amount: -billedSubtotal },
            { accountId: platformRevenueId, amount: -billedFee },
          ],
          userBalanceDelta: { userId: ownerId, amount: billedSubtotal },
        });
      } else {
        settlementTxnId = await postTransaction(client, {
          type: "job_escrow_release",
          referenceType: "job",
          referenceId: job.id,
          lines: [
            { accountId: platformEscrowId, amount: billedTotal },
            { accountId: platformRevenueId, amount: -billedTotal },
          ],
        });
      }
    }

    await client.query(
      `UPDATE jobs SET status = $1, completed_at = now(), settlement_transaction_id = $2,
         billed_subtotal_usd = $3, billed_fee_usd = $4, billed_total_usd = $5 WHERE id = $6`,
      [status, settlementTxnId, billedSubtotal.toFixed(2), billedFee.toFixed(2), billedTotal.toFixed(2), job.id]
    );

    await client.query("COMMIT");
    res.json({ ok: true });
  } catch (err) {
    await client.query("ROLLBACK");
    throw err;
  } finally {
    client.release();
  }
});
